delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuDeltaParquetFileFormatUtils.scala

```diff
@@ -0,0 +1,159 @@
+import org.apache.spark.sql.vectorized.{ColumnarBatch, ColumnVector}
+
+
+object GpuDeltaParquetFileFormatUtils {
+  /**
+   * Row number of the row in the file. When used with [[FILE_PATH_COL]] together, it can be used
+   * as unique id of a row in file. Currently to correctly calculate this, the caller needs to
+   * set both [[isSplitable]] to false, and [[RapidsConf.PARQUET_READER_TYPE]] to "PERFILE".
+   */
+  val METADATA_ROW_IDX_COL: String = "__metadata_row_index"
+  val METADATA_ROW_IDX_FIELD: StructField = StructField(METADATA_ROW_IDX_COL, LongType,
+    nullable = false)
+
+  val METADATA_ROW_DEL_COL: String = "__metadata_row_del"
+  val METADATA_ROW_DEL_FIELD: StructField = StructField(METADATA_ROW_DEL_COL, BooleanType,
+    nullable = false)
+
+
+  /**
+   * File path of the file that the row came from.
+   */
+  val FILE_PATH_COL: String = "_metadata_file_path"
+  val FILE_PATH_FIELD: StructField = StructField(FILE_PATH_COL, StringType, nullable = false)
+
+  /**
+   * Add a metadata column to the iterator. Currently only support [[METADATA_ROW_IDX_COL]].
+   */
+  def addMetadataColumnToIterator(
+      schema: StructType,
+      delVector: Option[Roaring64Bitmap],
+      input: Iterator[ColumnarBatch],
+      maxBatchSize: Int): Iterator[ColumnarBatch] = {
+    val metadataRowIndexCol = schema.fieldNames.indexOf(METADATA_ROW_IDX_COL)
+    val delRowIdx = schema.fieldNames.indexOf(METADATA_ROW_DEL_COL)
+    if (metadataRowIndexCol == -1 && delRowIdx == -1) {
+      return input
+    }
+    var rowIndex = 0L
+    input.map { batch =>
+      withResource(batch) { _ =>
+        val rowIdxCol = if (metadataRowIndexCol == -1) {
+          None
+        } else {
+          Some(metadataRowIndexCol)
+        }
+
+        val delRowIdx2 = if (delRowIdx == -1) {
+          None
+        } else {
+          Some(delRowIdx)
+        }
+        val newBatch = addMetadataColumns(rowIdxCol, delRowIdx2, delVector,maxBatchSize,
+          rowIndex, batch)
+        rowIndex += batch.numRows()
+        newBatch
+      }
+    }
+  }
+
+  private def addMetadataColumns(
+      rowIdxPos: Option[Int],
+      delRowIdx: Option[Int],
+      delVec: Option[Roaring64Bitmap],
+      maxBatchSize: Int,
+      rowIdxStart: Long,
+      batch: ColumnarBatch): ColumnarBatch = {
+    val rowIdxCol = rowIdxPos.map { _ =>
+      withResource(Scalar.fromLong(rowIdxStart)) { start =>
+        GpuColumnVector.from(CudfColumnVector.sequence(start, batch.numRows()),
+          METADATA_ROW_IDX_FIELD.dataType)
+      }
+    }
+
+    val delVecCol = delVec.map { delVec =>
+      withResource(Scalar.fromBool(false)) { s =>
+        withResource(CudfColumnVector.fromScalar(s, batch.numRows())) { c =>
+          var table = new Table(c)
+          val posIter = new RoaringBitmapIterator(
+            delVec.getLongIteratorFrom(rowIdxStart),
+            rowIdxStart,
+            rowIdxStart + batch.numRows(),
+          ).grouped(Math.min(maxBatchSize, batch.numRows()))
```

Comment:
Curious why we would iterate in smaller batches?  We're asking the GPU to do tiny amounts of work multiple times by using a small batch size here.  Not sure why this is configurable and set to a relatively small batch size.

Reply:
Understand the OOM concern, but 4096 still seems really small.  That's not a lot of data for the GPU to process per iteration, and it's only 32K on the CPU.